Source/WebDriver/WebSocketServer.cpp

```diff
@@ -0,0 +1,211 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebSocketServer.h"
+
+#include "CommandResult.h"
+#include "Session.h"
+#include "WebDriverService.h"
+#include <algorithm>
+#include <optional>
+#include <wtf/JSONValues.h>
+#include <wtf/UUID.h>
+#include <wtf/text/MakeString.h>
+#include <wtf/text/WTFString.h>
+
+namespace WebDriver {
+
+WebSocketServer::WebSocketServer(WebSocketMessageHandler& messageHandler, WebDriverService& service)
+    : m_messageHandler(messageHandler)
+    , m_service(service)
+{
+}
+
+void WebSocketServer::addStaticConnection(WebSocketMessageHandler::Connection&& connection)
+{
+    m_staticConnections.push_back(WTFMove(connection));
+}
+
+void WebSocketServer::addConnection(WebSocketMessageHandler::Connection&& connection, const String& sessionId)
+{
+    m_connectionToSession.add(WTFMove(connection), sessionId);
+}
+
+bool WebSocketServer::isStaticConnection(const WebSocketMessageHandler::Connection& connection)
+{
+    return std::count(m_staticConnections.begin(), m_staticConnections.end(), connection);
+}
+
+void WebSocketServer::removeStaticConnection(const WebSocketMessageHandler::Connection& connection)
+{
+    m_staticConnections.erase(std::find(m_staticConnections.begin(), m_staticConnections.end(), connection));
+}
+
+void WebSocketServer::removeConnection(const WebSocketMessageHandler::Connection& connection)
+{
+    auto it = m_connectionToSession.find(connection);
+    if (it != m_connectionToSession.end())
+        m_connectionToSession.remove(it);
+}
+
+RefPtr<Session> WebSocketServer::session(const WebSocketMessageHandler::Connection& connection)
+{
+    String sessionId;
+
+    for (const auto& pair : m_connectionToSession) {
+        if (pair.key == connection) {
+            sessionId = pair.value;
+            break;
+        }
+    }
+
+    if (sessionId.isNull())
+        return { };
+
+    auto existingSession = m_service.session();
```

Comment:
```suggestion
    auto& existingSession = m_service.session();
```

Reply:
Had to go with `const auto&`, as `m_service.session()` returns a rvalue and the compiler complained about assigning it to a `lvalue reference`.